app/(authenticated)/admin/users/UserView.tsx

```diff
@@ -0,0 +1,156 @@
+import { z } from "zod";
+import { useEffect, useState } from "react";
+import {
+  CountControls,
+  PageControls,
+  PaginationProvider,
+} from "@/components/Pagination";
+import { SearchBar } from "@/components/SearchBar";
+import { useValidSearchParams } from "@/lib/searchParams/validate";
+import { getSearchParamsString } from "@/lib/searchParams/util";
+import { UserCard } from "./UserCard";
+import { useQuery } from "@tanstack/react-query";
+import { fetchUsersAction, TFetchUsers } from "./actions";
+
+export function UserView(props: { initialUsers: TFetchUsers }) {
+  const pathname = usePathname();
+  const router = useRouter();
+
+  // Get and force validate search params
+  const rawSearchParams = useSearchParams();
+  const validSearchParams = useValidSearchParams(
+    searchParamsSchema,
+    rawSearchParams,
+  );
+
+  const usersQuery = useQuery({
+    initialData: props.initialUsers,
+    queryKey: ["admin:users", validSearchParams],
+    queryFn: async () => {
+      const res = await fetchUsersAction(validSearchParams);
+      if (!res.ok) {
+        throw new Error("An error occurred updating roles.");
+      } else {
+        return res;
+      }
+    },
+  });
+
+  const currentRange = useCurrentRange(usersQuery.data);
+
+  const [searchParamsState, setSearchParamsState] = useState(validSearchParams);
+
+  // Push search params changes to router on state change
+  useEffect(() => {
+    const newSearchParamsString = getSearchParamsString(searchParamsState);
+    if (
+      getSearchParamsString(Object.fromEntries(rawSearchParams.entries())) !=
+      newSearchParamsString
+    ) {
+      router.push(`${pathname}?${newSearchParamsString}`);
+    }
+  }, [searchParamsState]);
+
+  function updateState(state: Partial<z.infer<typeof searchParamsSchema>>) {
+    setSearchParamsState({
+      ...searchParamsState,
+      ...state,
+    });
+  }
+
+  if (!usersQuery.isSuccess) {
+    return <LoadingOverlay />;
+  }
+
+  // Wrapped in ScrollArea to avoid jerky scrolling on page change
+  return (
+    <ScrollArea>
+      <PaginationProvider
+        count={{
+          current: searchParamsState.count,
+          set(count) {
+            updateState({ count });
+          },
+          values: ["10", "25", "50", "100"],
+          range: currentRange,
+        }}
+        page={{
+          current: usersQuery.data.page,
+          set(page) {
+            updateState({ page });
+          },
+          total: Math.ceil(usersQuery.data.total / searchParamsState.count),
+        }}
+        totalItems={usersQuery.data.total}
+      >
+        <Stack>
+          <SearchBar
+            default={validSearchParams.query}
+            onChange={(query) => {
+              updateState({
+                query: query !== "" ? query : undefined,
+              });
+            }}
+            label="Search by Name or Email"
+            description="Search must start of the first, last or nick name, or username."
```

Comment:
```suggestion
            description="Search must start with the first, last, or nick name."
```

(rip usernames 2023-2024)